### PR TITLE
feat: new precompile secp256r1

### PR DESCRIPTION
**Pull request by umadayal**
## Overview
Includes precompile for secp256r1 along with a test in examples/patch-testing calling this precompile.

## Details

- Added curve Secp256r1 to enum CurveType in crates/curves/src/lib.rs
- Defined parameters such as curve modulus, a, b, and generator in crates/curves/src/weierstrass/secp256r1.rs
- Include new curve type to syscalls/precompiles/weierstrass/add, double and decompress
- Updated the decompress air so it can now be used for curves with non-zero a

With this new precompile, the p256 signature verification takes 55,581 cycles.

### CODE REVIEW

File: crates/core/executor/src/syscalls/code.rs [hunk 1]
```diff
@@ -169,7 +178,9 @@ impl SyscallCode {
             0x00_01_01_29 => SyscallCode::BN254_FP2_ADD,
             0x00_01_01_2A => SyscallCode::BN254_FP2_SUB,
             0x00_01_01_2B => SyscallCode::BN254_FP2_MUL,
-            0x00_00_01_1C => SyscallCode::BLS12381_DECOMPRESS,
```
Comment (kevjue) [inline crates/core/executor/src/syscalls/code.rs:172]: Is this remove intended?

Comment (umadayal) [reply crates/core/executor/src/syscalls/code.rs:172]: oh yes, thats an accident, thanks for catching it!

File: crates/core/executor/src/syscalls/precompiles/weierstrass/double.rs [hunk 2]
```diff
@@ -57,6 +62,6 @@ impl<E: EllipticCurve> Syscall for WeierstrassDoubleAssignSyscall<E> {
     }
 
     fn num_extra_cycles(&self) -> u32 {
-        0
```
Comment (kevjue) [inline crates/core/executor/src/syscalls/precompiles/weierstrass/double.rs:60]: I actually think that this should be reverted back to 0.

Based on what we found last night, the read and write of the memory slot is done within the same row, so no need to add an additional cycle.

Note that the opcode will need to be modified to have 0 extra cycles.

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        p256::AffinePoint::decompress(bytes_be.into(), Choice::from(sign as u8)).unwrap();
+    let point = computed_point.to_encoded_point(false);
+
+    let x = BigUint::from_bytes_be(point.x().unwrap());
+    let y = BigUint::from_bytes_be(point.y().unwrap());
+    AffinePoint::<E>::new(x, y)
+}
+
+pub fn secp256r1_sqrt(n: &BigUint) -> BigUint {
+    // println!("n: {}", n);
```
Comment (kevjue) [inline crates/curves/src/weierstrass/secp256r1.rs:114]: remove?

File: examples/Cargo.toml [hunk 4]
```diff
@@ -58,12 +58,16 @@ sp1-sdk = { path = "../crates/sdk" }
 sp1-lib = { path = "../crates/zkvm/lib", default-features = false }
 sp1-zkvm = { path = "../crates/zkvm/entrypoint", default-features = false }
 
+
```
Comment (kevjue) [inline examples/Cargo.toml:61]: remove.

File: examples/Cargo.toml [hunk 5]
```diff
@@ -58,12 +58,16 @@ sp1-sdk = { path = "../crates/sdk" }
...
 [patch.crates-io]
 curve25519-dalek = { git = "https://github.com/sp1-patches/curve25519-dalek", branch = "patch-curve25519-v4.1.3" }
 curve25519-dalek-ng = { git = "https://github.com/sp1-patches/curve25519-dalek-ng", branch = "patch-v4.1.1" }
 # ecdsa-core = { git = "https://github.com/sp1-patches/signatures", package = "ecdsa", branch = "patch-ecdsa-v0.16.8" }
 # Note: This branch of ecdsa-core points to SP1 branch ratan/impl-add-assign-fixes
-ecdsa-core = { git = "https://github.com/sp1-patches/signatures", package = "ecdsa", branch = "ratan/secp256k1-add-fixes-v0.16.8" }
+ecdsa-core = { git = "https://github.com/sp1-patches/signatures", package = "ecdsa", branch =  "umadayal/secp256r1" }
+
```
Comment (kevjue) [inline examples/Cargo.toml:68]: remove white space and comment.

File: examples/patch-testing/program/src/main.rs [hunk 6]
```diff
@@ -15,15 +15,17 @@ use ed25519_dalek::{
 
 use sha2_v0_10_6::{Digest as Digest_10_6, Sha256 as Sha256_10_6};
 // use sha2_v0_10_8::{Digest as Digest_10_8, Sha256 as Sha256_10_8};
-use sha2_v0_9_8::{Digest as Digest_9_8, Sha256 as Sha256_9_8};
-use tiny_keccak::{Hasher, Keccak};
+// use sha2_v0_9_8::{Digest as Digest_9_8, Sha256 as Sha256_9_8};
```
Comment (kevjue) [inline examples/patch-testing/program/src/main.rs:18]: revert change?

File: crates/core/executor/Cargo.toml [hunk 7]
```diff
@@ -43,7 +43,7 @@ vec_map = { version = "0.8.2", features = ["serde"] }
 enum-map = { version = "2.7.3", features = ["serde"] }
 
 [dev-dependencies]
-sp1-zkvm = { workspace = true }
+sp1-zkvm = { workspace = true, features = ["lib"] }
```
Comment (ratankaliani) [inline crates/core/executor/Cargo.toml:46]: Why do we need this feature?

File: crates/core/machine/src/operations/field/field_op.rs [hunk 8]
```diff
@@ -1,7 +1,8 @@
 use std::fmt::Debug;
 
 use crate::air::WordAirBuilder;
-use num::{BigUint, Zero};
+use num::BigUint;
+use num::Zero;
```
Comment (ratankaliani) [inline crates/core/machine/src/operations/field/field_op.rs:5]: nit: fix

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs [hunk 9]
```diff
@@ -107,20 +114,26 @@ impl<E: EllipticCurve + WeierstrassParameters> WeierstrassDecompressChip<E> {
...
 
         let sqrt_fn = match E::CURVE_TYPE {
             CurveType::Secp256k1 => secp256k1_sqrt,
+            CurveType::Secp256r1 => secp256r1_sqrt,
             CurveType::Bls12381 => bls12381_sqrt,
             _ => panic!("Unsupported curve"),
         };
-        let y = cols.y.populate(record, shard, &x_3_plus_b, sqrt_fn);
-
+        // let y = cols.y.populate(record, shard, &x_3_plus_b_plus_a_mul_x, sqrt_fn);
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs:135]: @umadayal Please go through this PR meticulously and catch all of the unneeded whitespace and leftover comments.

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs [hunk 10]
```diff
@@ -357,7 +379,8 @@ where
             local.is_real,
         );
 
-        local.y.eval(builder, &local.x_3_plus_b.result, local.y.lsb, local.is_real);
+        // local.y.eval(builder, &local.x_3_plus_b_plus_a_mul_x.result, local.y.lsb, local.is_real);
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs:382]: ^

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs [hunk 11]
```diff
@@ -573,7 +601,9 @@ mod tests {
             let public_key = secret_key.public_key();
             let encoded = public_key.to_encoded_point(false);
             let decompressed = encoded.as_bytes();
+            println!("decompressed: {:?}", decompressed);
             let compressed = public_key.to_sec1_bytes();
+            println!("compressed: {:?}", compressed);
```
Comment (ratankaliani) [inline crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs:606]: Remove comments

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs [hunk 12]
```diff
@@ -587,4 +617,35 @@ mod tests {
...
+
+        for _ in 0..num_tests {
+            let secret_key = p256::SecretKey::random(&mut rng);
+            let public_key = secret_key.public_key();
+            let encoded = public_key.to_encoded_point(false);
+            let decompressed = encoded.as_bytes();
+            let encoded_compressed = public_key.to_encoded_point(true);
+            println!("decompressed: {:?}", decompressed);
+            let compressed = encoded_compressed.as_bytes();
+            println!("compressed: {:?}", compressed);
```
Comment (ratankaliani) [inline crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_decompress.rs:637]: Remove comments

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_double.rs [hunk 13]
```diff
@@ -51,6 +51,7 @@ pub struct WeierstrassDoubleAssignCols<T, P: FieldParameters + NumWords> {
     pub p_ptr: T,
     pub p_access: GenericArray<MemoryWriteCols<T>, P::WordsCurvePoint>,
     pub(crate) slope_denominator: FieldOpCols<T, P>,
+    // pub(crate) slope_denominator_is_zero: IsZeroOperation<T>,
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_double.rs:54]: ^

File: crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_double.rs [hunk 14]
```diff
@@ -233,11 +236,22 @@ impl<F: PrimeField32, E: EllipticCurve + WeierstrassParameters> MachineAir<F>
         let mut values = zeroed_f_vec(num_rows * num_cols);
         let chunk_size = 64;
 
+        let num_words_field_element = E::BaseField::NB_LIMBS / 4;
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/core/machine/src/syscall/precompiles/weierstrass/weierstrass_double.rs:239]: Agreed, add a comment and replace with a constant.

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+            "48439561293906451759052585252797914202762949526041747995844080717082404635286",
+        )
+        .unwrap();
+        let y = BigUint::from_str(
+            "36134250956749795798585127919587881956611106672985015071877198253568414405109",
+        )
+        .unwrap();
+        (x, y)
+    }
+    //0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/curves/src/weierstrass/secp256r1.rs:80]: ^

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        let x = BigUint::from_str(
+            "48439561293906451759052585252797914202762949526041747995844080717082404635286",
+        )
+        .unwrap();
+        let y = BigUint::from_str(
+            "36134250956749795798585127919587881956611106672985015071877198253568414405109",
+        )
+        .unwrap();
+        (x, y)
+    }
```
Comment (ratankaliani) [inline crates/curves/src/weierstrass/secp256r1.rs:79]: Add a source for this above this fn

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+use typenum::{U32, U62};
+
+use super::{SwCurve, WeierstrassParameters};
+use crate::{
+    params::{FieldParameters, NumLimbs},
+    AffinePoint, CurveType, EllipticCurve, EllipticCurveParameters,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+/// Secp256k1 curve parameter
```
Comment (ratankaliani) [inline crates/curves/src/weierstrass/secp256r1.rs:24]: nit: Secp256r1

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 18]
```diff
@@ -0,0 +1,150 @@
+//! Modulo defining the Secp256r1 curve and its base field. The constants are all taken from
+//! https://en.bitcoin.it/wiki/Secp256k1.
+
+use std::str::FromStr;
+
+use elliptic_curve::{sec1::ToEncodedPoint, subtle::Choice};
+use generic_array::GenericArray;
+// use k256::{elliptic_curve::point::DecompressPoint, FieldElement};
```
Comment (ratankaliani) [inline crates/curves/src/weierstrass/secp256r1.rs:8]: nit: remove

File: crates/curves/src/weierstrass/secp256r1.rs [hunk 19]
```diff
@@ -0,0 +1,150 @@
+//! Modulo defining the Secp256r1 curve and its base field. The constants are all taken from
+//! https://en.bitcoin.it/wiki/Secp256k1.
```
Comment (ratankaliani) [inline crates/curves/src/weierstrass/secp256r1.rs:2]: nit: This isn't correct

**Review by ratankaliani — COMMENTED** [review]
It'd be great if you could add a PR description that specifies what your SP1 PR is doing.

Also, there's many places where there are outdated comments, and incorrect documentation (secp256r1 vs. secp256k1). Can you please meticulously go through and confirm that these are all correct. Especially given the scope of the PR, when there are small issues it makes it difficult to review.

**Review by kevjue — APPROVED** [review]
It's looking really good!  I'll approve, but please address a few related style and commenting related notes.

Also, please confirm with @tamirhemo about possibly merging the fields in the `decompress` table.

**Pull request merged**